controllers/configurationpolicy_controller.go

```diff
@@ -710,6 +714,145 @@ func currentlyUsingWatch(plc *policyv1.ConfigurationPolicy) bool {
 	return plc.Spec.EvaluationInterval.IsWatchForNonCompliant()
 }
 
+func getFormattedTemplateErr(err error) (complianceMsg string, formattedErr error) {
+	if errors.Is(err, templates.ErrInvalidAESKey) || errors.Is(err, templates.ErrAESKeyNotSet) {
+		return `The "policy-encryption-key" Secret contains an invalid AES key`, err
+	}
+
+	if errors.Is(err, templates.ErrInvalidIV) {
+		return fmt.Sprintf(
+			`The "%s" annotation value is not a valid initialization vector`, IVAnnotation,
+		), fmt.Errorf("%w: %w", ErrPolicyInvalid, err)
+	}
+
+	return err.Error(), err
+}
+
+func (r *ConfigurationPolicyReconciler) resolveObjectTemplatesRaw(
+	plc *policyv1.ConfigurationPolicy,
+	tmplResolver *templates.TemplateResolver,
+	resolveOptions *templates.ResolveOptions,
+) error {
+	objRawBytes := []byte(plc.Spec.ObjectTemplatesRaw)
+	plc.Spec.ObjectTemplates = []*policyv1.ObjectTemplate{}
+	resolveOptions.InputIsYAML = true
+
+	var err error
+
+	defer func() {
+		if err == nil {
+			return
+		}
+
+		complianceMsg, _ := getFormattedTemplateErr(err)
+
+		statusChanged := addConditionToStatus(plc, -1, false, "Error processing template", complianceMsg)
+		if statusChanged {
+			r.recordInfoEvent(plc, true)
+		}
+
+		r.updatedRelatedObjects(plc, []policyv1.RelatedObject{})
+
+		// Note: don't clean up child objects when there is a template violation
+
+		r.addForUpdate(plc, statusChanged)
+	}()
```

Comment:
There are only two returns currently in this function where this defer is useful (`err != nil`). Are there plans to have more? I remember wanting to refactor something like this a while ago, there were quite a few more cases at that point though. For only two (and two returns where `err == nil`), I'm not sure if the complexity is worth it. But if it matches another function with more cases, or there is a good possibility of additional errors in this function to be handled in the future, then this is great.

Reply:
Seems like there are fewer instances of `addConditionToStatus(plc, -1,` than I remember. Maybe they got cleaned up some other way at some point.

I could go either way on this defer.

Reply:
@JustinKuli I refactored this to remove the defer but still not duplicate the error handling. Let me know what you think!